climada/util/value_representation.py

```diff
@@ -179,3 +177,66 @@ def value_to_monetary_unit(values, n_sig_dig=None, abbreviations=None):
         mon_val = [sig_dig(val, n_sig_dig=n_sig_dig) for val in mon_val]
 
     return (mon_val, name)
+
+
+def safe_divide(numerator, denominator, replace_with=np.nan):
+    """
+    Safely divide two arrays or scalars.
+
+    This function handles division by zero and NaN values in the numerator or denominator.
+    If the division results in infinity or NaN, it is replaced by a specified value.
+
+    Parameters
+    ----------
+    numerator : array-like or scalar
+        The numerator for division.
+    denominator : array-like or scalar
+        The denominator for division. Division by zero is handled safely.
+    replace_with : float, optional
+        The value to use in place of division results that are infinity or NaN.
+        By default, it is NaN.
+
+    Returns
+    -------
+    array-like or scalar
+        The result of the division. If the division is not possible (e.g., division by zero),
+        it returns the value specified in `replace_with`.
+
+    Notes
+    -----
+    The function uses numpy's `true_divide` for array-like inputs and handles both
+    scalar and array-like inputs for the numerator and denominator. Division by zero
+    or NaN values in the input will result in the `replace_with` value in the output.
+
+    Examples
+    --------
+    >>> _safe_divide(1, 0)
+    nan
+
+    >>> _safe_divide(1, 0, replace_with=0)
+    0
+
+    >>> _safe_divide([1, 0, 3], [0, 0, 3])
+    array([nan, nan,  1.])
+    """
+    # Handle array inputs
+    if not np.isscalar(numerator) or not np.isscalar(denominator):
+        # If either input is an array and contains NaN, replace the whole result with replace_with
```

Comment:
seems counter intuitive to me. why not `safe_divide([nan, nan, 1.], 4) == [nan, nan, 0.25]` ?
I don't understand why division by zero is dealt with element wise but if any of the elements is `nan` all should turn to `nan`?
Since `safe_divide(4, 0)` is equal to `safe_divide(4, nan)`, i.m.o., `safe_divide([4, 4], [1, 0])` should also be equal to `safe_divide([4, 4], [1, nan])`.

Reply:
good point. the choice in the safe_divide function to treat NaN values in array inputs differently than division by zero is indeed not necessary. I updated the function:

`    with np.errstate(divide="ignore", invalid="ignore"):
        result = np.true_divide(numerator, denominator)
        
        # Check if the result is a scalar
        if np.isscalar(result):
            if not np.isfinite(result):
                return replace_with
        else:
            # Replace infinities, NaNs, and division by zeros in np.ndarray
            result[~np.isfinite(result)] = replace_with

    return result`